multicluster/controllers/multicluster/internal/remote_cluster.go

```diff
@@ -0,0 +1,320 @@
+		remoteClusterManager: remoteClusterManager,
+	}
+
+	err = remoteClusterManager.AddRemoteCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return cluster, nil
+}
+
+/**
+ * When a member is added to a ClusterSet, a specific ServiceAccount is created on the
+ * leader cluster which allows the member to access into the common area. This ServiceAccount
+ * has an associated Secret which must be copied into the Member Cluster as an opaque secret.
+ * Name of this secret is part of the ClusterSet spec for this member. This method reads
+ * the Secret given by that name.
+ */
+func getSecretCACrtAndToken(namespace string, secretName string) ([]byte, []byte, error) {
+	var err error
+	mutex.Lock()
+	if secretClient == nil {
+		secretClient, err = client.New(ctrl.GetConfigOrDie(), client.Options{})
+		if err != nil {
+			mutex.Unlock()
+			return nil, nil, err
+		}
+	}
+	mutex.Unlock()
+	secretObj := &v1.Secret{}
+	secretNamespacedName := types.NamespacedName{
+		Namespace: namespace,
+		Name:      secretName,
+	}
+	err = secretClient.Get(context.TODO(), secretNamespacedName, secretObj)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	caData, found := secretObj.Data[v1.ServiceAccountRootCAKey]
+	if !found {
+		return nil, nil, fmt.Errorf("ca.crt data not found in secret %v", secretName)
+	} else {
+	}
+
+	token, found := secretObj.Data[v1.ServiceAccountTokenKey]
+	if !found {
+		return nil, nil, fmt.Errorf("token not found in secret %v", secretName)
+	}
+
+	return caData, token, nil
+}
+
+func (r *remoteCluster) SendMemberAnnounce() error {
+	log := r.log.WithName("MemberAnnounce" + string(r.ClusterID))
+
+	log.Info("Writing member announce", "cluster", r.GetClusterID())
+
+	memberAnnounceList := &multiclusterv1alpha1.MemberClusterAnnounceList{}
+	if err := r.List(context.TODO(), memberAnnounceList, client.InNamespace(r.GetNamespace())); err != nil {
+		return err
+	}
+	var localClusterMemberAnnounce multiclusterv1alpha1.MemberClusterAnnounce
+	localClusterMemberAnnounceExist := false
+	if len(memberAnnounceList.Items) != 0 {
+		for _, memberAnnounce := range memberAnnounceList.Items {
+			if memberAnnounce.ClusterID == string(r.remoteClusterManager.GetLocalClusterID()) {
+				localClusterMemberAnnounceExist = true
+				localClusterMemberAnnounce = memberAnnounce
+				break
+			}
+		}
+	}
+	ctx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFunc()
+	if localClusterMemberAnnounceExist {
+		if r.remoteClusterManager.GetElectedLeaderClusterID() != common.INVALID_CLUSTER_ID {
+			localClusterMemberAnnounce.LeaderClusterID = string(r.remoteClusterManager.GetElectedLeaderClusterID())
+		}
+
+		if localClusterMemberAnnounce.Annotations == nil {
+			localClusterMemberAnnounce.Annotations = make(map[string]string)
+		}
+		localClusterMemberAnnounce.Annotations["touch-ts"] = time.Now().String()
+		// do an update
+		if err := r.Update(ctx, &localClusterMemberAnnounce, &client.UpdateOptions{}); err != nil {
```

Comment:
you should check existing memberannounce and do comparison before make a real update call, I see the existing memberannounce being updated every 5s in leader cluster with only `touch-ts` updated. btw, it's better to change `touch-ts` to a more clear name like `touch-timestamp`.

Reply:
I wanted to update it every interval. This way the leader can check if there are 3 consecutive missed updates, then, it will mark the cluster as disconnected. Because the leader cannot distinguish between when a member is disconnected vs, when it is just not writing member-announce because there is no change.